### PR TITLE
VACMS-10422 Add VBA fields to migration

### LINKED ISSUE #10422: CMS Add new VBA Facility fields to VBA Facility Migration
## Description

The fields added to the VBA Facility content type in #10421 need to also be added to the VBA facility migration. Use the VAMC Facility migration as a reference point for setting up source and processing details for each one. 

### CMS Team
Please check the team(s) that will do this work.

- [ ] `Program`
- [ ] `Platform CMS Team`
- [ ] `Sitewide Crew`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public Websites`
- [x] `⭐️ Facilities`
- [ ] `⭐️ User support`

### PR DESCRIPTION
**Pull request by omahane**
## Description

Closes #10422 

## Testing done

Manually

## Screenshots

A VBA Facility after migration on local dev
Screen Shot 2022-09-09 at 2 06 24 PM

## QA steps
Test the migration of the facility types currently using the `va_field_office_hours` plugin during migration

### VBA
As an admin user
1. Go to "/admin/structure/migrate/manage/facility/migrations/va_node_facility_vba/execute"
2. Choose Import (making sure to check the "Update" option)
3. Click "Execute"
   - Validate that 455 nodes were updated
4. Go to "/admin/content/"
5. Open 10 different VBA facilities from the first page. (The VBA Facilities should all be at the top (as the most recently updated)  
   - Validate that the data in each VBA facility corresponds with the data in the API (https://sandbox-api.va.gov/services/va_facilities/v0/facilities/all)
   - Validate that the hours in the CMS are in 12-hour format

### VAMC Facility
As an admin user
1. Go to "/admin/structure/migrate/manage/facility/migrations/va_node_health_care_local_facility/execute"
2. Choose Import
3. Click "Execute"
   - Validate that nodes were updated
4. Go to "/admin/content/"
5. Open 5 different VAMC facilities from the first page. (They should be at the top, as recently updated.)
  - Validate that the data in each VBA facility corresponds with the data in the API (https://sandbox-api.va.gov/services/va_facilities/v0/facilities/all)
   
### Vet Center
As an admin user
1. Go to "/admin/structure/migrate/manage/facility/migrations/va_node_facility_vet_centers/execute"
2. Choose Import
3. Click "Execute"
   - Validate that nodes were updated
4. Go to "/admin/content/"
5. Open 5 different Vet Center facilities from the first page. (They should be at the top, as recently updated.)
  - Validate that the data in each VBA facility corresponds with the data in the API (https://sandbox-api.va.gov/services/va_facilities/v0/facilities/all)
  

### Vet Center Mobile Vet Center
As an admin user
1. Go to "/admin/structure/migrate/manage/facility/migrations/va_node_facility_vet_centers_mvc/execute"
2. Choose Import
3. Click "Execute"
   - Validate that nodes were updated
4. Go to "/admin/content/"
5. Open 5 different Vet Center Mobile Vet Center facilities from the first page. (They should be at the top, as recently updated.)
  - Validate that the data in each VBA facility corresponds with the data in the API (https://sandbox-api.va.gov/services/va_facilities/v0/facilities/all)

### Vet Center Outstation
As an admin user
1. Go to "/admin/structure/migrate/manage/facility/migrations/va_node_facility_vet_centers_os/execute"
2. Choose Import
3. Click "Execute"
   - Validate that nodes were updated
4. Go to "/admin/content/"
5. Open 5 different Vet Center Outstation facilities from the first page. (They should be at the top, as recently updated.)
  - Validate that the data in each VBA facility corresponds with the data in the API (https://sandbox-api.va.gov/services/va_facilities/v0/facilities/all)

### Test the following pages against the API
- [Buffalo VA Regional Benefit Office](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/3955/edit)
  - vba_307 (has "By appointment only")
- [VetSuccess on Campus at Syracuse University](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/3957/edit)
  - vba_307b ("7:30a.m. - 4:00 p.m.")
- [Pre-Discharge Site at U.S. Army Garrison Humphreys](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/42635/edit)
  - vba_311a("8:00 a.m. -12:00 p.m.; 1:00 p.m.- 4:00 p.m.")
- [Pre-Discharge Site at Landstuhl Regional Medical Center](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/42636/edit)
  - vba_311b("9:00 a.m. -12:00 p.m.; 1:00 p.m.- 3:00 p.m.")
- [Atlanta Veteran Readiness and Employment Office](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4015/edit)
  - vba_316a("8:00 a.m - 4:00 p.m.")
- [VetSuccess on Campus at University of Florida](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4049/edit)
  - vba_317t("6:30 a.m.- 3:00 p.m.")
- [VetSuccess on Campus at Austin Peay State University](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4073/edit)
  - vba_320a("7:30 a.m. - 12:00 p.m.") - 12 pm
- [Pre-Discharge Site at Fort Knox](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4121/edit)
  - vba_327f("12:00 p.m. - 4:00 p.m.") - starts at 12 pm
- [Pre-Discharge Site at Davis-Monthan Air Force Base](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/42653/edit)
  - vba_345l("8:00 a.m -4:00 p.m.") - no space after hyphen
- [Boise VA Regional Benefits Office](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4233/edit)
  - vba_347("8:00AM-4:30PM")-upper case
- [Portland VA Regional Benefit Office](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4237/edit)
  - vba_348("8:00am - 4:00pm") - no space after minute
- [Portland Veteran Readiness and Employment Office](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4238/edit)
  - vba_348a("Closed 8:00am - 4:00pm") - closed or open these hours? // a number of 348* ones like this
- [VetSuccess on Campus at Tarrant County College Northeast Campus](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4267/edit)
  - vba_349u("8:00 a.m. to 4:30 p.m.") - has "to" instead of "-" 
- [Cheyenne VA Regional Benefit Office](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4338/edit)
  - vba_442("8:00 a.m. -3:30PM") - inconsistent am/pm and spacing
- [Pre-Discharge Site at Fort Wainwright Soldier for Life SFL Center](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/4355/edit)
  - vba_463c("08:30 AM – 4:00pm")- bad range character
- [Fort Harrison VA Medical Center](https://pr10679-lnfqr9ctulkx3loq5e6nqsp7igx2wsjz.ci.cms.va.gov/node/1080/edit)
  - vha_436("24/7")-not an actual range

  

### Definition of Done

- [ ] Documentation has been updated, if applicable.
- [ ] Tests have been added if necessary.
- [ ] Automated tests have passed.
- [ ] Code Quality Tests have passed.
- [ ] Acceptance Criteria in related issue are met.
- [ ] Manual Code Review Approved.

### Select Team for PR review

- [ ] `Platform CMS Team`
- [ ] `Sitewide program`
- [ ] `⭐️ Sitewide CMS`
- [ ] `⭐️ Public websites`
- [x] `⭐️ Facilities`
- [ ] `⭐️ User support`

### CMS user-facing annoucement

Is an announcement needed to let editors know of this change?
- [ ] Yes, and it's written in issue ____ and queued for publication.
  - [ ] Merge and ping the UX writer so they are ready to publish after deployment
- [ ] Yes, but it hasn't yet been written
  - [ ] Don't merge yet -- ping the UX writer to write and queue content
- [ ] No announcement is needed for this code change.
  - [ ] Merge & carry on unburdened by announcements

### CODE REVIEW

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php [hunk 1]
```diff
@@ -96,6 +99,31 @@ public function multiple() {
...
+ * @param string $hour_range
+ *   The string of opening and closing hours for a given day.
+ *
+ * @return string
+ *   A string with extraneous characters and letters capitalized.
+ */
+function normalize_hours($hour_range) {
+  // Set the replacement array with all characters to remove or make uppercase.
+  $replace = [
+    "/^0/"   => "",
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php:114]: Is this one needed? 
Using the line from clean_time()  it seems like it handles it accordingly for pm
image

and this code handles it for am
image

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php [hunk 2]
```diff
@@ -96,6 +99,31 @@ public function multiple() {
...
+    "/\-0/"  => "",
+    "/\./"   => "",
+    "/\s+/"  => "",
+    "/\:/"   => "",
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php:118]: I don't think this one is needed either.  I think it gets handled in cleantime()

image

Unless you are certain it is needed, we should remove it.  It is kind of a broad net that could alter a comment.

**Review by swirtSJW — CHANGES_REQUESTED** [review]
A couple thoughts in comments below.

Also as far as testing steps, you have done a good job accounting for what is expected.  However we haven't accounted for checking for the unexpected.  You are making changes to a process plugin that is used by other migrations that we know are migrating different formats of the data.  We want to make sure we didn't break that.  (please add testing steps to vailidate we did not break existing.

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php [hunk 3]
```diff
@@ -51,7 +51,10 @@ public function transform($value, MigrateExecutableInterface $migrate_executable
       'saturday' => '6',
       'sunday' => '0',
     ];
+
     foreach ($hours as $day => $hour) {
+      // Remove extra characters and make letters uppercase.
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php]: This has been removed.

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php [hunk 4]
```diff
@@ -96,6 +99,31 @@ public function multiple() {
...
+ *
+ * @return string
+ *   A string with extraneous characters and letters capitalized.
+ */
+function normalize_hours($hour_range) {
+  // Set the replacement array with all characters to remove or make uppercase.
+  $replace = [
+    "/^0/"   => "",
+    "/\-0/"  => "",
+    "/\./"   => "",
```
[the remark this answers is not in the document]
Comment (omahane) [reply docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php]: This has been replaced by only 2 items with optional periods.

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php (hunk 2 above)
Comment (omahane) [reply docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php]: Much of this is removed.

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php [hunk 5]
```diff
@@ -96,6 +101,27 @@ public function multiple() {
...
+ * @return string
+ *   A string with extraneous characters and letters capitalized.
+ */
+function normalize_hours($hour_range) {
+  // Set the replacement array with all characters to remove or make uppercase.
+  $replace = [
+    "/a\.?m\.?/i" => "AM",
+    "/p\.?m\.?/i" => "PM",
+    "/\–/"   => "-",
+    "/to/"   => "-",
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php:119]:
```suggestion
    "/\–/" => "-",
    "/to/" => "-",
```
I'm not sure how this is not getting codesniffer errors when you try to commit.  phpcs should be catching this.

File: docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php [hunk 6]
```diff
@@ -96,6 +101,27 @@ public function multiple() {
...
+ * @param string $hour_range
+ *   The string of opening and closing hours for a given day.
+ *
+ * @return string
+ *   A string with extraneous characters and letters capitalized.
+ */
+function normalize_hours($hour_range) {
+  // Set the replacement array with all characters to remove or make uppercase.
+  $replace = [
+    "/a\.?m\.?/i" => "AM",
```
Comment (swirtSJW) [inline docroot/modules/custom/va_gov_migrate/src/Plugin/migrate/process/VAFieldOfficeHours.php:116]: NICE

**Review by swirtSJW — CHANGES_REQUESTED** [review]
Looking good.  Only a CS issue.

**Comment by swirtSJW** [comment]
Migrations are running right now on the pr environment. 
Will QA shortly

**Comment by swirtSJW** [comment]
This is testing out great except for some non-usa outliers with less than truthful data 
image

**Comment by swirtSJW** [comment]
I think we should go ahead and merge this, and then I will raise this all in [#cms-lighthouse](https://dsva.slack.com/archives/C02BTJTDFTN) because there are 2 external tickets that would have to happen before we could properly implement a country fix on the CMS.

**Review by swirtSJW — APPROVED** [review]
All this checks out great.  other than the out of country issue with the data which is something that will need to be resolved elsewhere first.

Thank you for your fantastic QA steps.  Well done.

**Pull request merged**